WTrackMenu: Add ability to select loaded track in library

ninomp opened the pull request:
This PR allows user to highlight (select) track that is currently loaded to deck or sampler in library. This works across library features (Tracks, AutoDJ, Playlists, Crates).

ninomp committed:
WTrackMenu: Add ability to select/highlight loaded track in library

daschuer commented:
Nice. Thank you.

ninomp committed:
WTrackMenu: Add missing emit keyword

Russe commented:
This is a very nice feature, thank you.
I tested this PR with Utbuntu and it works like a charm.

One questions, do you think it's also possible to mark or highlight (not select) loaded tracks automatically in library?
This could be an additional feature.

uklotzde commented on src/widget/wtracktableview.h:
```diff
@@ -69,6 +69,10 @@ class WTrackTableView : public WLibraryTableView {
     bool slotRestoreCurrentViewState() {
         return restoreCurrentViewState();
     };
+    void slotSelectTrack(const TrackId& trackId) {
```
Please move the implementation into the .cpp file. I suggest to define only trivial, single-line functions inline.

uklotzde reviewed (approved):
One minor nitpick, otherwise fine.

ninomp committed:
WTrackTableView: Move implementation of slotSelectTrack() to .cpp file

daschuer commented:
Thank you for the feature. It works OK. 
Can you also scroll to the selected track if it is already selected. 
It would be also nice, if the current would be selected in case you have searched a similar track. 
The feature dos not work in case AutoDJ is selected. In this case the cover art is shown regardless of the track in in the AutoDJ List or not.
My expectation was that this feature works for the currently selected library table. Helpfull to see the current track in the context of the others in a crate or library.

daschuer commented on src/widget/wtracktableview.cpp:
```diff
@@ -1011,6 +1011,11 @@ void WTrackTableView::slotAddToAutoDJReplace() {
 }
 
+void WTrackTableView::slotSelectTrack(const TrackId& trackId) {
+    setCurrentTrackId(trackId);
```
I think here is a check missing if the track is in that Table at all.

ninomp committed:
WTrackTableView: Scroll to selected track

ninomp committed:
Merge commit 'd8826cab4d9155ab1ffd5dcc77b603b696292017' into select-in-library

ninomp committed:
Merge commit '6eb12c02d2e60e9128cd0d20672b489c3394c155' into select-in-library

ronso0 commented:
@ninomp Please reword the last two commits to provide meaningful messages, rebase onto the commit before [b80c7b3](https://github.com/mixxxdj/mixxx/pull/4688/commits/b80c7b3354c1cfdffce294230a97108807a5dc7f) and force-push (no one has reviewed them so far so it should be okay)

ronso0 commented:
> @ninomp Please reword the last two commits to provide meaningful messages, rebase onto the commit before [b80c7b3](https://github.com/mixxxdj/mixxx/pull/4688/commits/b80c7b3354c1cfdffce294230a97108807a5dc7f) and force-push (no one has reviewed them so far so it should be okay)

Sorry I didn't notice those were merges from main. Why didn't you use the usual `merge upstream/main`?

ronso0 commented on src/widget/wtrackproperty.cpp:
```diff
@@ -21,7 +21,8 @@ constexpr WTrackMenu::Features kTrackMenuFeatures =
         WTrackMenu::Feature::Properties |
-        WTrackMenu::Feature::UpdateReplayGainFromPregain;
+        WTrackMenu::Feature::UpdateReplayGainFromPregain |
+        WTrackMenu::Feature::SelectInLibrary;
```
This also needs to be added to WTrackWidgetGroup, used for the track color bar in the decks

ronso0 commented on src/widget/wtracktableview.cpp:
```diff
@@ -1034,6 +1039,11 @@ void WTrackTableView::slotAddToAutoDJReplace() {
 }
 
+void WTrackTableView::slotSelectTrack(const TrackId& trackId) {
+    setCurrentTrackId(trackId, 0, true);
+    setSelectedTracks({trackId});
```
```suggestion
    if (setCurrentTrackId(trackId, 0, true)) {
        setSelectedTracks({trackId});
    }
```

ronso0 commented on src/library/library.cpp:
```diff
@@ -400,6 +400,7 @@ void Library::bindLibraryWidget(
             &Library::restoreModelState,
             pTrackTableView,
             &WTrackTableView::slotRestoreCurrentViewState);
+    connect(this, &Library::selectTrack, pTrackTableView, &WTrackTableView::slotSelectTrack);
```
for readability, please break this line like the surrounding connections.

ronso0 commented:
re @daschuer's observation
> The feature dos not work in case AutoDJ is selected. In this case the cover art is shown regardless of the track in in the AutoDJ List or not.

This happens when clicking _any_ sidebar root item (feature root), except Tracks and History.
With those, the table model is simply not "shown" when the view is selected, so the `selectTrack` request goes to the previously selected WTrackTableView. Btw that also clears the previous track selection there :|

@ninomp
Since I messed with the library a few times already I was curious and took a look.
For features where the root view has a track view (AutoDJ, Analyze, Recording), I managed to fix it by adding `emit showTrackModel( [trackModel] );` to `[Feature]::activate`. If you mind that approach take a look at 8cfd397773
For other root views _without_ a tracks table (Crates, Playlists, Browse, some external libraries?), no idea..

Maybe the `selectTrack` signal should rather be connected in `Library::addFeature`, and then each feature needs a `slotSelectTrack` slot? Maybe also make use of `WLibrary::getActiveView()`, or add `bool `WLibrary::isCurrentViewTracksTable()`? Possible, if features store the WLibrary pointer in `feature::bindLibraryWidget`.
https://github.com/mixxxdj/mixxx/blob/6f8676073e1689a1cc749a701d909570d0cafb0e/src/widget/wlibrary.cpp#L91-L93

Sorry, I'm aware this may be overwhelming since your current minimal approach is actually beautifully simple and seemed to work fine, too.

one more thing... :grimacing: 
I'd actually expect the "Select track" action to be greyed out if the current library view is not a WTrackTableView, though it is probably complete overkill to make the decks' track menu aware of the tracks table (if possible at all).

ronso0 committed:
Library: emit showTrackmodel when activating feature root with tracks table

This makes the track menu 'Select track in library' action work in feature root views that
have a tracks table (AutoDJ, Analyze), though it's not working in Browse and Recording
because the BrowseModel does currently not support getTrackRows().

Issue still unresolved:
When triggering this section while a feature root view with no tracks table is active
the track will be selected in the previously selected tracks table.

ronso0 committed:
DlgRecording: remove irrelevant forward declaration

ronso0 committed:
WTrackMenu: disable 'Select track...' if track can't be selected

ronso0 committed:
WLibrary/WTrackTableView: fix & tweak some debug messages

daschuer commented:
https://github.com/mixxxdj/mixxx/pull/4740 has been merged, so this can be closed. @ninomp Thank you for starting this nice feature.
